dora/core/common/src/main/java/alluxio/membership/EtcdMembershipManager.java

```diff
@@ -116,84 +115,76 @@ public void join(WorkerInfo workerInfo) throws IOException {
   }
 
   @Override
-  public List<WorkerInfo> getAllMembers() throws IOException {
-    List<WorkerServiceEntity> registeredWorkers = retrieveFullMembers();
-    return registeredWorkers.stream()
-        .map(e -> new WorkerInfo().setIdentity(e.getIdentity())
-            .setAddress(e.getWorkerNetAddress()))
-        .collect(Collectors.toList());
+  public WorkerClusterView getAllMembers() throws IOException {
+    Iterable<WorkerInfo> workerInfoIterable = parseWorkersFromEtcdKvPairs(
+        mAlluxioEtcdClient.getChildren(getRingPathPrefix()))
+        .map(w -> new WorkerInfo()
+            .setIdentity(w.getIdentity())
+            .setAddress(w.getWorkerNetAddress()))
+        ::iterator;
+    return new WorkerClusterView(workerInfoIterable);
   }
 
-  private List<WorkerServiceEntity> retrieveFullMembers() throws IOException {
-    List<WorkerServiceEntity> fullMembers = new ArrayList<>();
-    List<KeyValue> childrenKvs = mAlluxioEtcdClient.getChildren(getRingPathPrefix());
-    for (KeyValue kv : childrenKvs) {
-      try {
-        WorkerServiceEntity entity = new WorkerServiceEntity();
-        entity.deserialize(kv.getValue().getBytes());
-        fullMembers.add(entity);
-      } catch (JsonParseException ex) {
-        // Ignore
-      }
-    }
-    return fullMembers;
+  @Override
+  public WorkerClusterView getLiveMembers() throws IOException {
+    Iterable<WorkerInfo> workerInfoIterable = parseWorkersFromEtcdKvPairs(
+        mAlluxioEtcdClient.mServiceDiscovery.getAllLiveServices())
+        .map(w -> new WorkerInfo()
+            .setIdentity(w.getIdentity())
+            .setAddress(w.getWorkerNetAddress()))
+        ::iterator;
+    return new WorkerClusterView(workerInfoIterable);
```

Comment:
i imagine the interface is a little easier if you can
```suggestion
  @Override
  public WorkerClusterView getLiveMembers() throws IOException {
    return getAllMembers().getLiveWorkers();
```

but maybe that's not so easy because it needs a 
```
class WorkerClusterView {
  public WorkerClusterView getLiveWorkers();
}
```

Reply:
Take a look at the implementation for failed workers. You need to first get a set of live worker IDs, and exclude those workers from a list of all workers, to get a list of failed workers. Two etcd queries are involved in this process/